langs/tree-sitter-langs-build.el

```diff
@@ -20,6 +20,12 @@
   (require 'pcase)
   (require 'cl-lib))
 
+(defcustom tree-sitter-langs-bin-dir (expand-file-name ".cache/tsc-langs-bin/"
```

Comment:
It should be `tree-sitter-langs`, not `tsc-langs`. The latter implies this belongs to the core lib `tsc`.

However, I think it's better to keep the default value the same, i.e. 

```emacs-lisp
(file-name-as-directory
 (concat (file-name-directory (locate-library "tree-sitter-langs.el")) "bin"))
```

Reply:
The reason is that I'm now not entirely convinced that the grammar binaries should be under a data/cache directory, since they are actually *code* (shared libs). Furthermore, the current code already adds `.el` to the filename when calling `locate-library`, which seems to work in gccemacs.

It's good to have this knob, but I don't think it should be turned by default.

Reply:
I think the problem is when upgrading `tree-sitter-langs` package (once this is on MELPA or using Quelpa). On Windows, since the grammar binaries is still loaded, we cannot delete the old package and it will most likely make the upgrading fails.

Since the grammar binaries has different version than `tree-sitter-langs`, I think it would make sense to have it store under data/cache directory. WDYT?

Reply:
> I think the problem is when upgrading `tree-sitter-langs` package (once this is on MELPA or using Quelpa). On Windows, since the grammar binaries is still loaded, we cannot delete the old package and it will most likely make the upgrading fails.

Putting the binaries under a fixed directory under `~/.emacs.d` doesn't solve this, since `tree-sitter-langs-install-grammars` will try to overwrite them, which will be an error on Windows. In fact, I don't think the existing implementation has this problem. Since the directory is in the package's installation directory, upgrading the package will result in a new directory, and `package.el` doesn't delete the old version of a package when upgrading IIRC.

Properly solving the upgrading issue probably requires the binary directory to be tied to the bundle version. We would also need to tell user to restart Emacs for it to actually work.

Reply:
> Putting the binaries under a fixed directory under `~/.emacs.d` doesn't solve this, since `tree-sitter-langs-install-grammars` will try to overwrite them, which will be an error on Windows.

> Properly solving the upgrading issue probably requires the binary directory to be tied to the bundle version. We would also need to tell user to restart Emacs for it to actually work.

I think current code is checking for `BUNDLE_VERSIOIN` before overwrite the binary, so normally it would be fine.
In case of binary is updating, the overwrite will fail, so wouldn't that also halt the new version being recorded in `tree-sitter-langs--bundle-version-file`?
User will see that the updating of binary failed, but it will likely being reinstalled again after Emacs is restarted.
This is of course not the properly way to solve this problem, but it works and no temp file is left around.



> Since the directory is in the package's installation directory, upgrading the package will result in a new directory, and `package.el` doesn't delete the old version of a package when upgrading IIRC.

When I try to upgrade from `list-packages`, the old packages are deleted by default. What happens is the new package is marked as `I` and old package is marked as `D` when update interactive, so old package will be deleted.

Since the grammar binaries are update much less than the Elisp package itself, provide binaries in different location would make the upgrade process fail much less (and user need to restart Emacs less), wouldn't it?

Anw, I've change back to default location for grammar binaries, I can custom that out in my `init.el` instead.